pymatviz/ptable.py

```diff
@@ -255,9 +193,77 @@ def data_preprocessor(data: SupportedDataType) -> pd.DataFrame:
             vmin: 1.0
             vmax: 12.0
     """
+
+    def format_pd_dataframe(df: pd.DataFrame) -> pd.DataFrame:
+        """Fix pd.DataFrame that does not meet expected format."""
+
+        def fix_df_elem_as_col(df: pd.DataFrame) -> pd.DataFrame:
+            """Fix pd.DataFrame where elements are in a single column."""
+            # Copy and reset index to move element names to a column
+            new_df = df.copy()
+            new_df = new_df.reset_index()
+
+            # Find the column with element names
+            elem_col_name = None
+            for col in new_df.columns:
+                if set(new_df[col]).issubset(set(map(str, Element))):
+                    elem_col_name = col
+                    break
+
+            # Fix failed: cannot find the elements column
+            if elem_col_name is None:
+                return None
+
+            # Rename the column with elements and set it as index
+            new_df = new_df.rename(columns={elem_col_name: Key.element})
+            new_df = new_df.set_index(Key.element)
+
+            # Zip the remaining values into a single column
+            value_cols = [
+                col_name for col_name in new_df.columns if col_name != Key.element
+            ]
+            new_df[Key.heat_val] = new_df[value_cols].apply(list, axis=1)
+
+            # Drop the old value columns
+            return new_df[[Key.heat_val]]
+
+        # Check if input DataFrame is in expected format
+        if Key.element == df.index.name and Key.heat_val in df.columns:
+            return df
+
+        # Re-format it to expected
+        warnings.warn("pd.DataFrame has unexpected format, trying to fix.")
+
+        # Try to search for elements as a column
+        fixed_df = fix_df_elem_as_col(df)
+        if fixed_df is not None:
+            return fixed_df
+
+        # Try to search for elements as a row
+        fixed_df = fix_df_elem_as_col(df.transpose())
+        if fixed_df is not None:
+            return fixed_df
+
+        raise RuntimeError("Cannot fix provided DataFrame.")
+
+    def handle_missing_and_anomaly(
```

Comment:
what's left to be done here?

Reply:
Good catch, it was not finished when we first refactor the ptable plotter with `PTableProjector`, and I have been procrastinating on this :( But I would do it soon :)

We would need to implement infinity and `NaN` handling here, and also need to update the unit test for this.

Reply:
we don't need to tackle it in this PR. let me know if you prefer to do this later

Reply:
Well... Let's handle this later then, for two reasons:
One, this PR is already pretty huge.

More importantly, I current don't have a solid (elegant) solution in my mind, on how to maintain the shape of nested arrays. If we were to fix bad values in the provided `DataFrame`, we would need to:
- First, explode (flatten) the `DataFrame` to check for NaN or infinity.
- If found, we would need to replace such values, but this seems a bit tricky as our supported data types are too flexible (and unpredictable).

For data without nested structure (`ptable_hists` and `ptable_heatmap/heatmap_splits`), this is fairly easy, we could simply record its original shape, and recover it:https://github.com/janosh/pymatviz/blob/5dd53ef31e3f1ef13aede234ef379d2eb6966e69/tests/test_ptable.py#L46-L53

But for plotters with deeper nested structure, this sounds a bit messy, for example `ptable_scatters`: https://github.com/janosh/pymatviz/blob/5dd53ef31e3f1ef13aede234ef379d2eb6966e69/tests/test_ptable.py#L576-L580

We could certain record its shape recursively, but I'm hoping that we could have a cleaner solution. Also during such handling, we also need to group data (i.e. we have to separate data for `x`/`y` and so on, instead of using the global average to infer a missing data for `y`).